app/assets/javascripts/discourse/app/initializers/subscribe-user-notifications.js

```diff
@@ -92,22 +92,19 @@ export default {
               );
             }
 
-            for (let idx = 0; idx < data.recent.length; idx++) {
-              let old;
-              while ((old = oldNotifications[idx])) {
-                const info = data.recent[idx];
-
-                if (old.get("id") !== info[0]) {
-                  oldNotifications.removeAt(idx);
-                } else {
-                  if (old.get("read") !== info[1]) {
-                    old.set("read", info[1]);
-                  }
-                  break;
-                }
-              }
-              if (!old) {
-                break;
+            const read = {};
+            data.recent.forEach((notification) => {
+              read[notification[0]] = notification[1];
+            });
+
+            // remove stale notifications and update existing ones
+            for (let idx = 0; idx < oldNotifications.length; ) {
+              const notification = oldNotifications[idx];
+              if (read[notification.id] === undefined) {
+                oldNotifications.removeAt(idx);
+              } else {
+                notification.set("read", read[notification.id]);
+                ++idx;
               }
```

Comment:
Could this non-`for(;;)` version work?

```suggestion
            // remove stale notifications and update existing ones
            const newNotifications = oldNotifications.map((notification) => {
              if (read[notification.id] !== undefined) {
                notification.set("read", read[notification.id]);
                return notification;
              }
            }).filter(Boolean);

            stale.results.set("content", newNotifications);
```